### ISSUE TITLE
Add licenses to the formats in KSY repo

### ISSUE THREAD

**Issue by KOLANICH** [kaitai-io/kaitai_struct#264]
Not all the formats in the repo have license metadata property. You can either add it or apply some license by default for the files **in the ```kaitai-io/kaitai_struct_formats``` repo** where license is not mentioned

**Comment by GreyCat** [comment]
Yeah, it's a known problem. Lots of my older specs miss them and, unfortunately, sometimes I forgot to ask the contributors to add it. We can't decide for them, so we should work towards getting this clear.

**Comment by GreyCat** [comment]
Checklist to fix / resolve:

- [ ] ipv4_packet: no license
- [ ] icmp_packet: no license
- [ ] ethernet_frame: no license
- [ ] ipv6_packet: no license
- [ ] tcp_segment: no license
- [ ] udp_datagram: no license
- [ ] dns_packet: no license
- [ ] microsoft_network_monitor_v2: no license
- [ ] dex: license "CC-BY-SA-3.0" does not look valid
- [ ] swf: no license
- [ ] python_pyc_27: no license
- [ ] mach_o: no license
- [ ] microsoft_pe: no license
- [ ] exif_be: no license
- [ ] psx_tim: no license
- [ ] exif: no license
- [ ] jpeg: no license
- [ ] png: no license
- [ ] wmf: no license
- [ ] exif_le: no license
- [ ] wav: license "BSD-3-Clause-Attribution" does not look valid
- [ ] stl: no license
- [ ] creative_voice_file: no license
- [ ] standard_midi_file: no license
- [ ] quicktime_mov: no license
- [ ] dbf: no license
- [ ] rar: no license
- [ ] nt_mdt: license "GPL-3.0+" does not look valid
- [ ] glibc_utmp: no license
- [ ] asn1_der: no license
- [ ] microsoft_cfb: no license
- [ ] vfat: no license
- [ ] lvm2: license "GFDL-1.3+" does not look valid
- [ ] iso9660: no license
- [ ] mbr_partition_table: no license
- [ ] ext2: no license
- [ ] ozf2: no license
- [ ] ftl_dat: no license
- [ ] quake_mdl: no license
- [ ] heroes_of_might_and_magic_agg: no license
- [ ] warcraft_2_pud: no license
- [ ] renderware_binary_stream: no license
- [ ] heroes_of_might_and_magic_bmp: no license
- [ ] quake_pak: no license
- [ ] windows_shell_items: no license
- [ ] regf: no license

**Comment by GreyCat** [comment]
A big question about adding licensing info goes to:

* @adventureloop — udp
* @LogicAndTrick — quake_mdl
* @koczkatamas — regf, mach_o, python_pyc_27
* @mnakamura1337 — renderware_binary_stream

If you don't want / can't commit it yourself, you can just state it here, I'll add it for you.

**Comment by LogicAndTrick** [comment]
I'm fine with using CC0 for quake_mdl.

**Comment by koczkatamas** [comment]
I am also fine with CC0, but I need to check first if I copied any licensed material (eg. as comment or doc field) by mistake.

**Comment by KOLANICH** [comment]
>GPL-3.0+ doesn't look valid?

[Really?](https://spdx.org/licenses/GPL-3.0+.html)

**Comment by koczkatamas** [comment]
Oh about `dex`: I used `CC-BY-SA-3.0` because [the original documentation the ksy based on](https://source.android.com/devices/tech/dalvik/dex-format) was licensed with `CC-BY-SA-3.0`.

Should we change this? Or remove the documentation parts which were copied from there?

**Comment by GreyCat** [comment]
@KOLANICH `GPL-3.0+` is actually a [license expression](https://spdx.org/spdx-specification-21-web-version#h.jxpfx0ykyb60), not just a license identifier, as it is expected by current parser (and, anyway, even `GPL-3.0` is not whitelisted there). Probably we need to fix the parser.

**Comment by GreyCat** [comment]
@koczkatamas dex is ok, it's just that `CC-BY-SA-3.0` should be whitelisted in the parser.

**Comment by adventureloop** [comment]
@GreyCat CC0 is fine

**Comment by GreyCat** [comment]
@adventureloop Thanks for a swift reply! Will add it now :)

**Comment by KOLANICH** [comment]
Also there are very slicky questions. We need to know answers to them (=consult a good lawer) and put them into Readme.md

Let we have some spec. It is copyrighted by corporations, it is explicitly written there. Is the ksy file (or any other product) based on this spec copyighted by the spec copyright holders? Does it mean that everyone wanting to develop low level things for PC platform is legally required to have business aggreement with Intel? Is the ksy file with docs borrowed from the spec  copyrighted by the spec holders? Is the ksy file reverse-engineered from open source software with viral license legally required to share the same license? I mean is it legal to create a ksy based on a spec and put a public domain-like license there (I use Unlicense)? If it is not legal, how about putting doc strings into a separate file licensed under a proprietary license, and putting the description into a file ticensed under public-domain like license and making KSC link the 2 files together when generating the code?

**Comment by Glitch0011** [comment]
Hi @GreyCat feel free to license the dns_packet as CC0, hell, you did as much work on it as I did with your comments if I remember right.

**Comment by GreyCat** [comment]
@Glitch0011 Thanks!

**Comment by GreyCat** [comment]
> Also there are very slick questions. We need to know answers to them (=consult a good lawer) and put them into Readme.md

I believe that, unfortunately, it won't be that simple. There are several major reasons for that: all copyright terms are slightly different, there are multiple countries involved which have somewhat different copyright laws, etc, etc. There would be no "100% guaranteed" rules. Even if we'll seek lawyer's advice, it would be for a particular case, not en masse.

> Is the ksy file with docs borrowed from the spec copyrighted by the spec holders?

It's not "copyrighted by the spec holders", it is definitely infringing someone's copyright, if there were no explicit permissions given that one can copy the text from a spec, and it's substantial enough to not be treated as "fair use". One the other hand, even if you don't copy the text directly, there's a still a chance of getting a lawsuit (and even winning one — see, for example, the infamous [Whelan](http://digital-law-online.info/lpdi1.0/treatise21.html) US decision — although in modern practice usually an [abstaction-filtration-comparison test](https://en.wikipedia.org/wiki/Abstraction-Filtration-Comparison_test) takes precedence after [CA vs Altai case](https://en.wikipedia.org/wiki/Computer_Associates_International,_Inc._v._Altai,_Inc.)).

> Is the ksy file reverse-engineered from open source software with viral license legally required to share the same license?

This is open source software, and it's *much* easier just to ping relevant author directly and ask, if in doubt. That's good in many aspects: (1) people getting praise and recognition for their past work they deserve, (2) people feeling that it was not in vain, (3) people getting to know about Kaitai Struct formats project (and may be they would want to contribute and use it themselves), (4) getting their definitive opinion is a 99.9% guarantee that no surprise lawsuits would be filed.

> I mean is it legal to create a ksy based on a spec and put a public domain-like license there?

When a corporation publishes a spec, it is usually done for some purpose, and it sometimes even outlines what they seem to be appropriate to do with it. Creation of software that *uses* that spec is usually legal, but direct copying of a spec text is usually not — although, again, all that stuff should be decided on case-by-case basis.

> (I use Unlicense)

May be that's not a very good idea — at least [FSF, OSI and Fedora project lawyers](https://en.wikipedia.org/wiki/Unlicense#Reception) recommend using CC0 over Unlicense.

> If it is not legal, how about putting doc strings into a separate file licensed under a proprietary license, and putting the description into a file ticensed under public-domain like license and making KSC link the 2 files together when generating the code?

I heavily doubt that would work. You're still copying and distributing slightly modified version of copyright-protected text that you don't have the right to distribute. One may invent crazier variants of your idea, like specifying text positions to copy from a verbatim spec file — i.e., literally, like:

```yaml
- id: foo
  type: u2
  doc:
    file: some_crazy_format_spec.html
    pos-start: 1000
    pos-end:  1080
```

but I doubt that anyone would use such an atrocity :) And, anyway, even that's still not 100% guarantee.

**Comment by KOLANICH** [comment]
>One may invent crazier variants of your idea, like specifying text positions to copy from a verbatim spec file
<details>
<summary>It's a very good idea</summary>
1 transform a pdf into a plain text
2 find the corresponding substrings in plain text by offsets
3 find these strings in pdf
4 navigate to that text in a pdf viewer
</details>
